File: lc/tree/ruby/balanced_binary_tree.rb
def is_balanced(root)
    return true unless root

    is_left_balanced = is_balanced root.left
    is_right_balanced = is_balanced root.right

    left_depth = tree_depth root.left
    right_depth = tree_depth root.right

    (left_depth - right_depth).abs <= 1 && is_left_balanced && is_right_balanced
end

private
def tree_depth(node)
    return 0 unless node

    left_depth = tree_depth node.left
    right_depth = tree_depth node.right

    [left_depth, right_depth].max + 1
end

File: lc/tree/java/BalancedBinaryTree.java
package lc.tree.java;

public class BalancedBinaryTree {
    public boolean isBalanced(TreeNode root) {
        if(root == null) return true;

        boolean isLeftBalanced = isBalanced(root.left);
        boolean isRightBalanced = isBalanced(root.right);

        int leftDepth = treeDepth(root.left);
        int rightDepth = treeDepth(root.right);

        return Math.max(leftDepth - rightDepth) <= 1 && isLeftBalanced && isRightBalanced;
    }

    private int treeDepth(TreeNode node) {
        if(node == null) return 0;

        int leftDepth = treeDepth(node.left);
        int rightDepth = treeDepth(node.right);

        return Math.max(leftDepth, rightDepth) + 1;
    }
}